File: laundry-reservation/python/src/laundry_reservation/reservation.py
from __future__ import annotations

from datetime import datetime

from .customer import Customer


class Reservation:
    def __init__(
        self,
        id: str,
        slot: datetime,
        machine_number: int,
        customer: Customer,
        pin: int,
    ) -> None:
        self._id = id
        self._slot = slot
        self._machine_number = machine_number
        self._customer = customer
        self._pin = pin
        self._is_used = False

    @property
    def id(self) -> str:
        return self._id

    @property
    def slot(self) -> datetime:
        return self._slot

    @property
    def machine_number(self) -> int:
        return self._machine_number

    @property
    def customer(self) -> Customer:
        return self._customer

    @property
    def pin(self) -> int:
        return self._pin

    @property
    def is_used(self) -> bool:
        return self._is_used

    def mark_used(self) -> None:
        self._is_used = True

    def update_pin(self, new_pin: int) -> None:
        self._pin = new_pin

File: laundry-reservation/csharp/src/LaundryReservation/Reservation.cs
namespace LaundryReservation;

public class Reservation
{
    public Guid Id { get; }
    public DateTime Slot { get; }
    public int MachineNumber { get; }
    public Customer Customer { get; }
    public int Pin { get; private set; }
    public bool IsUsed { get; private set; }

    public Reservation(Guid id, DateTime slot, int machineNumber, Customer customer, int pin)
    {
        Id = id;
        Slot = slot;
        MachineNumber = machineNumber;
        Customer = customer;
        Pin = pin;
    }

    public void MarkUsed() => IsUsed = true;

    public void UpdatePin(int newPin) => Pin = newPin;
}
